### PR TITLE
refactor(@angular/build): Auto-CSP support as an index file transformation.

### PR DESCRIPTION
**Pull request by aaronshim**
Auto-CSP is a feature to rewrite the `<script>` tags in a index.html file to either hash their contents or rewrite them as a dynamic loader script that can be hashed. These hashes will be placed in a CSP inside a `<meta>` tag inside the `<head>` of the document to ensure that the scripts running on the page are those known during the compile-time of the client-side rendered application.

This is an application of the recommended [Strict CSP](https://web.dev/articles/strict-csp#choose-hash) approach that is also recommended by the [OWASP CSP Cheat Sheet](https://cheatsheetseries.owasp.org/cheatsheets/Content_Security_Policy_Cheat_Sheet.html#strict-csp) and inspired by https://github.com/google/strict-csp. The approach of having build tooling auto-generate a hash-based CSP addresses the main concern around hash-based CSPs that generating and updating the hashes manually is difficult and error-prone.

## PR Checklist

Please check to confirm your PR fulfills the following requirements:

- [X] The commit message follows our guidelines: https://github.com/angular/angular-cli/blob/main/CONTRIBUTING.md#-commit-message-guidelines
- [X] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [X] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [X] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## What is the current behavior?

There is currently no logic to process the `index.html` to auto-generate hash-based CSPs.

## What is the new behavior?

Adding logic to process the `index.html` to auto-generate hash-based CSPs.

## Does this PR introduce a breaking change?

- [ ] Yes
- [X] No

### CODE REVIEW

File: packages/angular/build/src/utils/index-file/auto-csp.ts [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+    enableTrustedTypes?: boolean;
+    enableUnsafeEval?: boolean;
+  } = {
+    enableBrowserFallbacks: true,
+    enableTrustedTypes: false,
+    enableUnsafeEval: false,
+  },
+): string {
+  hashes = hashes || [];
+  let strictCspTemplate = {
```
Comment (dgp1130) [inline packages/angular/build/src/utils/index-file/auto-csp.ts:169]: Consider: I usually recommend `Map<string, T>` over `Record<string, T>` for anything with dynamically assigned keys to mitigate potential prototype pollution issues.

All the keys are string literals, so I don't think it's too important here, up to you if it's worth the effort.

File: packages/angular/build/src/utils/index-file/auto-csp.ts [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+/**
+ * Returns JS code for dynamically loading sourced (external) scripts.
+ * @param srcList A list of paths for scripts that should be loaded.
+ */
+function createLoaderScript(
+  srcList: SrcScriptTag[],
+  enableTrustedTypes = false,
+): string | undefined {
+  if (!srcList.length) {
+    return undefined;
```
Comment (dgp1130) [inline packages/angular/build/src/utils/index-file/auto-csp.ts:224]: Suggestion: Throw an error instead of returning `undefined`.

It's a bit confusing to have a `createLoaderScript` function that sometimes doesn't create a loader script. We already check `if (scriptContent)` above, so I think we should always have items in the list. If we ever don't have items we might not be enabling auto CSP as expected. I suspect it would be better to fail up front rather than no-op in that situation.

Removing `| undefined` from the type also makes usage of this function a little simpler.

File: packages/angular/build/src/utils/index-file/auto-csp.ts [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+/**
+ * The hash function to use for hash directives to use in the CSP.
+ */
+const HASH_FUNCTION = 'sha256';
+
+/**
+ * Store the appropriate attributes of a sourced script tag to generate the loader script.
+ */
+interface SrcScriptTag {
+  scriptType: 'src';
```
Comment (dgp1130) [inline packages/angular/build/src/utils/index-file/auto-csp.ts:22]: Question: Do we / will we need to check this type anywhere? Should we drop it?

File: packages/angular/build/src/utils/index-file/auto-csp.ts [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+ */
+function createLoaderScript(
+  srcList: SrcScriptTag[],
+  enableTrustedTypes = false,
+): string | undefined {
+  if (!srcList.length) {
+    return undefined;
+  }
+  const srcListFormatted = srcList
+    .map((s) => `['${s.src}',${s.type ? "'" + s.type + "'" : undefined}, ${s.async}]`)
```
Comment (dgp1130) [inline packages/angular/build/src/utils/index-file/auto-csp.ts:227]: Nit: Missing space between `'${s.src}',${s.type ...}`

Comment (dgp1130) [inline packages/angular/build/src/utils/index-file/auto-csp.ts:227]: Question: Do we need to escape `src` / `type` / `async`? Those could contain `"><script src="...">` which we probably wouldn't want (admittedly auto CSP should solve the script injection issue and exploiting this would require compromising source code / a dependency, but still something to potentially mitigate).

File: packages/angular/build/src/utils/index-file/auto-csp.ts [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+  });\n`
+    : `
+  var scripts = [${srcListFormatted}];
+  scripts.forEach(function(scriptUrl) {
+    var s = document.createElement('script');
+    s.src = scriptUrl[0];
+    s.type = scriptUrl[1];
+    s.async = !!scriptUrl[2];
+    document.body.appendChild(s);
+  });\n`;
```
Comment (dgp1130) [inline packages/angular/build/src/utils/index-file/auto-csp.ts:251]: Consider: You should be fine to use modern features like `const`, `() => {}`, `for-of`, destructuring, etc. Angular already limits support to browsers with those features. Not sure if you were avoiding those for compatibility or performance reasons.

File: packages/angular/build/src/utils/index-file/auto-csp_spec.ts [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+        </head>
+        <body>
+          <script>console.log('foo');</script>
+          <div>Some text </div>
+        </body>
+      </html>
+    `);
+
+    expect(result).toContain(
+      `<meta http-equiv="Content-Security-Policy" content="script-src 'strict-dynamic' 'sha256-1kOLrDKT3TBiHLcnxiGsc7HF/lyVJKLhoZDSn0UwCfo=' https: 'unsafe-inline';object-src 'none';base-uri 'self';">`,
```
Comment (dgp1130) [inline packages/angular/build/src/utils/index-file/auto-csp_spec.ts:25]: Suggestion: Rather than hard-coding the hashes which don't mean anything to humans anyways, could we replace this with something like:

```typescript
export(result).toContain(`... '${hash('console.log(\'foo\');')}' ...`);
```

I get that this runs the risk that the `hash` function is wrong, but I think that's outweighed by making the tests easier to read and avoid hard-coding or having to update hashes manually.

File: packages/angular/build/src/utils/index-file/auto-csp_spec.ts [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+        </body>
+      </html>
+    `);
+
+    expect(result).toContain(
+      `<meta http-equiv="Content-Security-Policy" content="script-src 'strict-dynamic' 'sha256-vkcRYaaRLTkuunplxAyjuivyXpK+pbbEfJB92l8u+aY=' https: 'unsafe-inline';object-src 'none';base-uri 'self';">`,
+    );
+    expect(result).toContain(
+      `var scripts = [['./main1.js',undefined, false],['./main2.js',undefined, false],['./main3.js','module', false]];`,
+    );
```
Comment (dgp1130) [inline packages/angular/build/src/utils/index-file/auto-csp_spec.ts:87]: Consider: Maybe verify that we only have one `<script>` tag in the output? That way we ensure they were batched together as expected.

**Review by dgp1130 — COMMENTED** [review]
Thanks for putting this together! Very excited to see this feature land.

I left a bunch of readability / refactoring suggestions, but none of them are really critical and what you have is pretty reasonable so feel free to decline any which aren't worth applying.

For the commit message, we should use `refactor` here since it doesn't actually expose any functionality yet. When you get to the "keystone" commit which exposes the relevant option for end users, that's where you would put `feat` to be included in the changelog.

File: packages/angular/build/src/utils/index-file/auto-csp.ts [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+
+/**
+ * Store the appropriate attributes of a sourced script tag to generate the loader script.
+ */
+interface SrcScriptTag {
+  scriptType: 'src';
+  src: string;
+  type?: string;
+  async?: boolean;
+}
```
[the remark this answers is not in the document]
Comment (aaronshim) [reply packages/angular/build/src/utils/index-file/auto-csp.ts:26]: Admittedly, I haven't done a lot of testing with the `defer` attribute, but for now, we can add it to the list of attributes that we want to pass along-- that way, worst case scenario is that we are left with some vestigial code.

File: packages/angular/build/src/utils/index-file/auto-csp.ts [hunk 9]
```diff
@@ -0,0 +1,252 @@
...
+function getScriptAttributeValue(tag: StartTag, attrName: string): string | undefined {
+  return tag.attrs.find((attr) => attr.name === attrName)?.value;
+}
+
+/**
+ * Calculates a CSP compatible hash of an inline script.
+ * @param scriptText Text between opening and closing script tag. Has to
+ *     include whitespaces and newlines!
+ */
+function hashInlineScript(scriptText: string): string {
```
[the remark this answers is not in the document]
Comment (aaronshim) [reply packages/angular/build/src/utils/index-file/auto-csp.ts:43]: We meant "inline script" as in a script tag with text content for us to hash-- renaming!

File: packages/angular/build/src/utils/index-file/auto-csp.ts [hunk 10]
```diff
@@ -0,0 +1,252 @@
...
+  const hash = crypto.createHash(HASH_FUNCTION).update(scriptText, 'utf-8').digest('base64');
+  return `'${HASH_FUNCTION}-${hash}'`;
+}
+
+/**
+ * Finds all `<script>` tags and creates a dynamic script loading block for consecutive `<script>` with `src` attributes.
+ * Hashes all scripts, both inline and generated dynamic script loading blocks.
+ * Inserts a `<meta>` tag at the end of the `<head>` of the document with the generated hash-based CSP.
+ *
+ * @param html Markup that should be processed.
```
[the remark this answers is not in the document]
Comment (aaronshim) [reply packages/angular/build/src/utils/index-file/auto-csp.ts:53]: Added!

File: packages/angular/build/src/utils/index-file/auto-csp.ts [hunk 11]
```diff
@@ -0,0 +1,252 @@
...
+ */
+export async function autoCsp(html: string): Promise<string> {
+  const { rewriter, transformedContent } = await htmlRewritingStream(html);
+
+  let openedScriptTag: StartTag | undefined = undefined;
+  let scriptContent: SrcScriptTag[] | undefined = undefined;
```
[the remark this answers is not in the document]
Comment (aaronshim) [reply packages/angular/build/src/utils/index-file/auto-csp.ts:59]: Done!

File: packages/angular/build/src/utils/index-file/auto-csp.ts [hunk 12]
```diff
@@ -0,0 +1,252 @@
...
+    }
+    rewriter.emitText(tag);
+  });
+
+  rewriter.on('endTag', (tag, html) => {
+    if (tag.tagName === 'script') {
+      const src = getScriptAttributeValue(openedScriptTag!, 'src');
+      openedScriptTag = undefined;
+
+      if (src) {
```
[the remark this answers is not in the document]
Comment (aaronshim) [reply packages/angular/build/src/utils/index-file/auto-csp.ts:105]: Added, thanks!

File: packages/angular/build/src/utils/index-file/auto-csp.ts [hunk 13]
```diff
@@ -0,0 +1,252 @@
...
+        scriptContent.push({
+          scriptType: 'src',
+          src: src,
+          type: getScriptAttributeValue(tag, 'type'),
+          async: !!getScriptAttributeValue(tag, 'async'),
+        });
+        return; // Skip writing my script tag until we've read it all...
+      }
+    }
+    // We are encountering the first start tag that's not <script src="..."> after a string of those.
```
[the remark this answers is not in the document]
Comment (aaronshim) [reply packages/angular/build/src/utils/index-file/auto-csp.ts:81]: Added, thanks!

File: packages/angular/build/src/utils/index-file/auto-csp.ts [hunk 14]
```diff
@@ -0,0 +1,252 @@
...
+    if (tag.tagName === 'body' || tag.tagName === 'html') {
+      // Write the loader script if a string of <script>s were the last opening tag of the document.
+      if (scriptContent) {
+        const loaderScript = createLoaderScript(scriptContent);
+        if (loaderScript) {
+          hashes.push(hashInlineScript(loaderScript));
+          rewriter.emitRaw(`<script>${loaderScript}</script>`);
+        }
+        scriptContent = undefined;
+      }
```
[the remark this answers is not in the document]
Comment (aaronshim) [reply packages/angular/build/src/utils/index-file/auto-csp.ts:119]: We'll opt for the first approach, and if the logic still looks weird, we can try the second later, thanks!

File: packages/angular/build/src/utils/index-file/auto-csp.ts [hunk 15]
```diff
@@ -0,0 +1,252 @@
...
+    }
+    rewriter.emitEndTag(tag);
+  });
+
+  const rewritten = await transformedContent();
+
+  // Second pass to add the header
+  const secondPass = await htmlRewritingStream(rewritten);
+  secondPass.rewriter.on('endTag', (tag, _) => {
```
[the remark this answers is not in the document]
Comment (aaronshim) [reply packages/angular/build/src/utils/index-file/auto-csp.ts:128]: Sure, I think that's easy enough to change for now!

File: packages/angular/build/src/utils/index-file/auto-csp.ts [hunk 16]
```diff
@@ -0,0 +1,252 @@
...
+        `<meta http-equiv="Content-Security-Policy" content="${getStrictCsp(hashes)}">`,
+      );
+    }
+    secondPass.rewriter.emitEndTag(tag);
+  });
+  return secondPass.transformedContent();
+}
+
+/**
+ * Returns a strict Content Security Policy for mittigating XSS.
```
[the remark this answers is not in the document]
Comment (aaronshim) [reply packages/angular/build/src/utils/index-file/auto-csp.ts:141]: Done!

File: packages/angular/build/src/utils/index-file/auto-csp.ts [hunk 17]
```diff
@@ -0,0 +1,252 @@
...
+    }
+  }
+
+  // If enabled, dangerous DOM sinks will only accept typed objects instead of
+  // strings.
+  if (cspOptions.enableTrustedTypes) {
+    strictCspTemplate = {
+      ...strictCspTemplate,
+      ...{ 'require-trusted-types-for': [`'script'`] },
+    };
```
[the remark this answers is not in the document]
Comment (aaronshim) [reply packages/angular/build/src/utils/index-file/auto-csp.ts:199]: Changed, thanks!

File: packages/angular/build/src/utils/index-file/auto-csp.ts [hunk 18]
```diff
@@ -0,0 +1,252 @@
...
+  // and 'strict-dynamic'.
+  if (cspOptions.enableBrowserFallbacks) {
+    // Fallback for Safari. All modern browsers supporting strict-dynamic will
+    // ignore the 'https:' fallback.
+    strictCspTemplate['script-src'].push('https:');
+    // 'unsafe-inline' is only ignored in presence of a hash or nonce.
+    if (hashes.length > 0) {
+      strictCspTemplate['script-src'].push(`'unsafe-inline'`);
+    }
+  }
```
[the remark this answers is not in the document]
Comment (aaronshim) [reply packages/angular/build/src/utils/index-file/auto-csp.ts:191]: For what it's worth, this is largely from https://github.com/google/strict-csp/blob/main/strict-csp/index.ts

File: packages/angular/build/src/utils/index-file/auto-csp.ts (hunk 2 above)
Comment (aaronshim) [reply packages/angular/build/src/utils/index-file/auto-csp.ts:224]: Done, thanks!

File: packages/angular/build/src/utils/index-file/auto-csp.ts (hunk 4 above)
Comment (aaronshim) [reply packages/angular/build/src/utils/index-file/auto-csp.ts:227]: Done.

File: packages/angular/build/src/utils/index-file/auto-csp.ts (hunk 3 above)
Comment (aaronshim) [reply packages/angular/build/src/utils/index-file/auto-csp.ts:22]: I'm not quite sure what you mean-- do you mean getting rid of this type definition entirely?

File: packages/angular/build/src/utils/index-file/auto-csp.ts (hunk 4 above)
Comment (aaronshim) [reply packages/angular/build/src/utils/index-file/auto-csp.ts:227]: Done

Comment (aaronshim) [reply packages/angular/build/src/utils/index-file/auto-csp.ts:227]: Attempted escaping without a dependency-- is it likely to break some behavior, in your opinion?

File: packages/angular/build/src/utils/index-file/auto-csp_spec.ts (hunk 6 above)
Comment (aaronshim) [reply packages/angular/build/src/utils/index-file/auto-csp_spec.ts:25]: Done!

File: packages/angular/build/src/utils/index-file/auto-csp_spec.ts (hunk 7 above)
Comment (aaronshim) [reply packages/angular/build/src/utils/index-file/auto-csp_spec.ts:87]: Tested!

File: packages/angular/build/src/utils/index-file/auto-csp.ts (hunk 1 above)
Comment (aaronshim) [reply packages/angular/build/src/utils/index-file/auto-csp.ts:169]: I'm not quite sure I'm following-- this is for `strictCspTemplate`?

File: packages/angular/build/src/utils/index-file/auto-csp.ts (hunk 4 above)
Comment (dgp1130) [reply packages/angular/build/src/utils/index-file/auto-csp.ts]: Hmm, this seems like a surprisingly tricky problem. We're trying to move a string from an HTML context to a JS context without altering its content, yet the two languages have different escaping rules. I don't think `encodeURI` is the right mechanism, it doesn't drop quotes.

```typescript
encodeURI("foo'bar"); // foo'bar
```

There's two sides to this. First, I think we need to escape the JS side (special characters in a single quoted string literal) such as `/foo'bar\'baz`:
1.  `'`, since that started the string literal and is the only token which can end it.
2. `\`, since that can escape other characters which might not have been intended to be escaped.

```typescript
function escape(str: string): string {
  // Convert `\` -> `\\` and `'` -> `\'`.
  // Probably can't use `replaceAll` just yet, I think you need a regex match with the 'g' flag?
  return str.replaceAll('\\', '\\\\').replaceAll("'", "\\'");
}

const srcListFormatted = srcList.map((s) => `['${escape(s.src)}', ...]`);
// ...
```

Second, we need to handle HTML entities such as `<script src="/foo&amp;bar">` which requests `/foo&bar`. I wonder if the HTML parser does this automatically. If you give it `<script src="/foo&amp;bar">`, does it give you a JS string containing `/foo&bar`? I feel like that would be the correct behavior for this parser and would generate a JS string literal of `'/foo&bar'` in the loader which seems like what we want?

Tests for all these cases would be super useful to make sure the HTML string and the generated JS string are equivalent. The ones which come to mind are:
* `/one'two\'three\\'four\\\'five` - Check quotes and backslashes.
* `/one&two&amp;three&amp;amp;four` - Check HTML entities.

There are other HTML entities like `#32;`, but I think if the named entities work we're probably fine? Not sure if there would be value in a test case there. Any other edge case scenarios we need to handle here?

File: packages/angular/build/src/utils/index-file/auto-csp.ts [hunk 19]
```diff
@@ -0,0 +1,322 @@
...
+ */
+function emitLoaderScript(
+  scriptContent: SrcScriptTag[],
+  hashes: string[],
+  rewriter: RewritingStream,
+) {
+  const loaderScript = createLoaderScript(scriptContent);
+  hashes.push(hashScriptText(loaderScript));
+  rewriter.emitRaw(`<script>${loaderScript}</script>`);
+}
```
[the remark this answers is not in the document]
Comment (aaronshim) [reply packages/angular/build/src/utils/index-file/auto-csp.ts:116]: You're so right, we are in JS so we can use all the language niceties like closure capture! 🎉

File: packages/angular/build/src/utils/index-file/auto-csp.ts [hunk 20]
```diff
@@ -0,0 +1,322 @@
...
+  let openedScriptTag: StartTag | undefined = undefined;
+  let scriptContent: SrcScriptTag[] = [];
+  let hashes: string[] = [];
+
+  rewriter.on('startTag', (tag, html) => {
+    if (tag.tagName === 'script') {
+      openedScriptTag = tag;
+      const src = getScriptAttributeValue(tag, 'src');
+
+      if (src) {
```
[the remark this answers is not in the document]
Comment (aaronshim) [reply packages/angular/build/src/utils/index-file/auto-csp.ts:138]: Perhaps I am misunderstanding-- but if we do `if (!src) return;` here, they won't get to later on:

```js
if (scriptContent.length > 0) {
  emitLoaderScript();
}
rewriter.emitStartTag(tag);
```

which we want script tags that *do not* have a src to do? (Emit a loader script based on the consecutive `<script src="...">` seen so far and then continue writing as normal)

Sorry, the logic is really tricky here and involves falling through various if statements-- but this is the rewritten version because before in the version where I had more if's and else-if's to denote each of the subcases distinctly, there was a lot of repetition.

File: packages/angular/build/src/utils/index-file/auto-csp.ts [hunk 21]
```diff
@@ -0,0 +1,322 @@
...
+            src: src,
+            type: scriptType,
+            async: !(getScriptAttributeValue(tag, 'async') === undefined),
+            defer: !(getScriptAttributeValue(tag, 'defer') === undefined),
+          });
+          return; // Skip writing my script tag until we've read it all.
+        }
+      }
+    }
+    // We are encountering the first start tag that's not <script src="..."> after a string of
```
[the remark this answers is not in the document]
Comment (aaronshim) [reply packages/angular/build/src/utils/index-file/auto-csp.ts:153]: I think we want the loader script to be written before a `<script>console.log(...)</script>`, if the previous tags were a string of `<script src=...>`, right?

This is captured in the unit test that does

```html
<script>console.log('foo');</script>
<script src="./main.js"></script>
<script src="./main2.js"></script>
<script>console.log('bar');</script>
<script src="./main3.js"></script>
<script src="./main4.js"></script>
<div>Some text </div>
```

Which I thought should be

```html
<script>console.log('foo');</script>
<script>...Loader script for main.js and main2.js</script>
<script>console.log('bar');</script>
<script>...Loader script for main3.js and main4.js</script>
<div>Some text </div>
```

As per a script with a malformed type (i.e. `<script type="not/js" ...>`), the current behavior is "leave it alone", since it's not guaranteed to run anyway according to the spec? If we want to be more forgiving, we can always rewrite it into a loader script if it also contains a `src` attribute, but I thought at least leaving it alone was preferred to silently stripping it from the HTML. (Regardless, we need the logic to fall through to `rewriter.emitStartTag` as written to preserve this behavior.)

File: packages/angular/build/src/utils/index-file/auto-csp.ts [hunk 22]
```diff
@@ -0,0 +1,322 @@
...
+    if (scriptContent.length > 0) {
+      emitLoaderScript(scriptContent, hashes, rewriter);
+      scriptContent = [];
+    }
+    rewriter.emitStartTag(tag);
+  });
+
+  rewriter.on('text', (tag, html) => {
+    if (openedScriptTag && !getScriptAttributeValue(openedScriptTag, 'src')) {
+      hashes.push(hashScriptText(html));
```
[the remark this answers is not in the document]
Comment (aaronshim) [reply packages/angular/build/src/utils/index-file/auto-csp.ts:168]: Done, thanks!

File: packages/angular/build/src/utils/index-file/auto-csp.ts (hunk 5 above)
Comment (aaronshim) [reply packages/angular/build/src/utils/index-file/auto-csp.ts:321]: I was doing so to be extra sure about correctness since our reference implementation (which we have been using for the testing and feasibility explorations so far) has been doing-- is the modern syntax something that you feel strongly about swapping out for the initial release or is there time to follow up in a later fix? Thanks!

File: packages/angular/build/src/utils/index-file/auto-csp_spec.ts [hunk 23]
```diff
@@ -0,0 +1,124 @@
...
+        </head>
+        <body>
+          <script src="./main.js"></script>
+          <div>Some text </div>
+        </body>
+      </html>
+    `);
+
+    expect(result).toContain(
+      `<meta http-equiv="Content-Security-Policy" content="script-src 'strict-dynamic' 'sha256-cfa69N/DhgtxzDzIHo+IFj9KPigQLDJgb6ZGZa3g5Cs=' https: 'unsafe-inline';object-src 'none';base-uri 'self';">`,
```
[the remark this answers is not in the document]
Comment (aaronshim) [reply packages/angular/build/src/utils/index-file/auto-csp_spec.ts:42]: Done, thanks for the suggestion!

File: packages/angular/build/src/utils/index-file/auto-csp.ts (hunk 4 above)
Comment (aaronshim) [reply packages/angular/build/src/utils/index-file/auto-csp.ts]: So, first of all, I really am enjoying this conversation, but for the sake of the threat model of whether our dynamic loader script would cause insecurities-- I don't think our CSP is meant to be a defense against *malicious developers*, especially with this flow, since all of the malicious payloads has to be around at compile time. (We imagine that if they have influence over the `src` attribute of a script tag in code that appears at compile time, they don't need to use this pathway to do a lot more harm.)

That being said, I think it's a great goal to make sure that developers can't inadvertently break our loader script! I don't think we need to worry about `.type` since that's fixed to one of the values allowed by `shouldDynamicallyLoadScriptTagBasedOnType`, but `.src` can be any free-form value.

I think we can stop the `'` case breaking out of our JS strings and rendering the JS unusable (by `replaceAll`ing the single-quotes) but I think there are other trivial JS escapes, like `</script>` appearing anywhere in the script to immediately exit the JS context back into the HTML context-- is this another case worth covering for? (Is there a developer that would set `<script src='./</script>'>` legitimately?) If so, I think `encodeURI` might be encoding `<` and `>`?

I added all of these as test cases. Thanks!

File: packages/angular/build/src/utils/index-file/auto-csp.ts (hunk 5 above)
Comment (dgp1130) [reply packages/angular/build/src/utils/index-file/auto-csp.ts:310]: What you have is fine, if you'd rather stick with that to stay aligned with your prior testing I'm ok with that. I just wasn't sure why you were using legacy syntax and wanted to point out that's not a hard constraint here.

+1 to fixing later if you'd rather go down that approach.

File: packages/angular/build/src/utils/index-file/auto-csp.ts [hunk 24]
```diff
@@ -0,0 +1,311 @@
...
+ * @param srcList A list of paths for scripts that should be loaded.
+ */
+function createLoaderScript(srcList: SrcScriptTag[], enableTrustedTypes = false): string {
+  if (!srcList.length) {
+    throw new Error('Cannot create a loader script with no scripts to load.');
+  }
+  const srcListFormatted = srcList
+    .map(
+      (s) =>
+        `['${encodeURI(s.src).replaceAll("'", "\\'")}', ${s.type ? "'" + encodeURI(s.type) + "'" : undefined}, ${s.async ? 'true' : 'false'}, ${s.defer ? 'true' : 'false'}]`,
```
Comment (dgp1130) [inline packages/angular/build/src/utils/index-file/auto-csp.ts:283]: Consider: `type` isn't a URI and it's limited to known values. I feel like we don't need `encodeURI` for it?

File: packages/angular/build/src/utils/index-file/auto-csp_spec.ts [hunk 25]
```diff
@@ -0,0 +1,170 @@
...
+    `);
+
+    const csps = getCsps(result);
+    expect(csps.length).toBe(1);
+    expect(csps[0]).toMatch(ONE_HASH_CSP);
+    // &amp; encodes correctly
+    expect(result).toContain(`'/foo&bar'`);
+    // Impossible to escape a string and create invalid loader JS with a '
+    // (Quotes and backslashes work)
+    expect(result).toContain(`'/one\\'two%5C\\'three%5C%5C\\'four%5C%5C%5C\\'five'`);
```
Comment (dgp1130) [inline packages/angular/build/src/utils/index-file/auto-csp_spec.ts:129]: Question: Why do we want to `encodeURI` and convert `\` to `%5C`? If I do `<script src="/foo\bar">` and `<script src="/foo%5Cbar">` the browser seems to request two different URLs. I feel like we'd want to just propagate the `\` character?

**Review by dgp1130 — APPROVED** [review]
Thanks for working through these comments! I don't have any significant concerns at this point, @clydin can you take a quick look too?

File: packages/angular/build/src/utils/index-file/auto-csp.ts (hunk 4 above)
Comment (dgp1130) [reply packages/angular/build/src/utils/index-file/auto-csp.ts]: Ooh, `</script>` is a really tricky one...

Yeah, I'm less concerned from the security angle, if an attacker compromised source code they could do a lot more damage than this particular attribute. It's not even that vulnerable to a compromised dependency since the `index.html` is a hard-coded static file, so even then we're very unlikely to parse dependency code here.

I'm mainly focused on correctness here. The user provided a `<script src="...">` string, and that string should be unchanged in the final `<script src="...">` we append to the DOM. It's probably a bit overly pedantic, I can't imagine any legitimate use cases for `</script>` in a URL for example and maybe it's not worth stressing about that. I could see URLs with a quote or `&amp;` though and would want to stay consistent. We can always fix up any missed edge cases later though, so I wouldn't get hung up on the specifics too much more than I already made you. :sweat_smile:

**Comment by dgp1130** [comment]
FYI: For a review, the commits you have are probably fine, but once this is approved and ready to submit, you'll need to squash/amend them to match the commit message conventions so we can merge it. Just a heads up.

**Comment by aaronshim** [comment]
> FYI: For a review, the commits you have are probably fine, but once this is approved and ready to submit, you'll need to squash/amend them to match the commit message conventions so we can merge it. Just a heads up.

For sure! I realized that the GitHub code review interface was hiding comments as "outdated" and such when I amended and force pushed on the same commit, so I decided to stop doing that until we are ready to merge!

File: packages/angular/build/src/utils/index-file/auto-csp.ts (hunk 24 above)
Comment (aaronshim) [reply packages/angular/build/src/utils/index-file/auto-csp.ts:283]: Agreed!

File: packages/angular/build/src/utils/index-file/auto-csp_spec.ts (hunk 25 above)
Comment (aaronshim) [reply packages/angular/build/src/utils/index-file/auto-csp_spec.ts:129]: That is... certainly unfortunate, I am perplexed why this is the case? I was under the impression that `encodeURI` was spec-compliant to only encode characters that would be treated the same.

What are our options here? If we move away from encodeURI, I am not sure what clean options we have that are along the lines of "escape everything... except for %5C".

Comment (aaronshim) [reply packages/angular/build/src/utils/index-file/auto-csp_spec.ts:129]: I think we are going to live with backslashes being `%5C` for now, as it is RFC 3986 compliant.

File: packages/angular/build/src/utils/index-file/auto-csp.ts [hunk 26]
```diff
@@ -0,0 +1,314 @@
...
+      const src = getScriptAttributeValue(tag, 'src');
+
+      if (src) {
+        // If there are any interesting attributes, note them down.
+        const scriptType = getScriptAttributeValue(tag, 'type');
+        if (shouldDynamicallyLoadScriptTagBasedOnType(scriptType)) {
+          scriptContent.push({
+            src: src,
+            type: scriptType,
+            async: !(getScriptAttributeValue(tag, 'async') === undefined),
```
Comment (alan-agius4) [inline packages/angular/build/src/utils/index-file/auto-csp.ts:137]:
```suggestion
            async: getScriptAttributeValue(tag, 'async') !== undefined,
```

File: packages/angular/build/src/utils/index-file/auto-csp.ts [hunk 27]
```diff
@@ -0,0 +1,314 @@
...
+            defer: !(getScriptAttributeValue(tag, 'defer') === undefined),
```
Comment (alan-agius4) [inline packages/angular/build/src/utils/index-file/auto-csp.ts:138]:
```suggestion
            defer: getScriptAttributeValue(tag, 'defer') !== undefined,
```

File: package.json [hunk 28]
```diff
@@ -174,6 +174,7 @@
     "ora": "5.4.1",
     "pacote": "19.0.0",
     "parse5-html-rewriting-stream": "7.0.0",
+    "parse5-sax-parser": "7.0.0",
```
[the remark this answers is not in the document]
Comment (clydin) [reply package.json:177]: Actually, opened up #28666 so once in you can just import from `./html-rewriting-stream`

File: packages/angular/build/src/utils/index-file/auto-csp_spec.ts (hunk 25 above)
Comment (dgp1130) [reply packages/angular/build/src/utils/index-file/auto-csp_spec.ts:129]: I'm ok with `%5C`, I'm inclined to go with that for now and revisit if we discover any problems.

File: package.json (hunk 28 above)
[the remark this answers is not in the document]
Comment (aaronshim) [reply package.json:177]: Done, thanks!

File: packages/angular/build/src/utils/index-file/auto-csp.ts (hunk 21 above)
[the remark this answers is not in the document]
Comment (aaronshim) [reply packages/angular/build/src/utils/index-file/auto-csp.ts:119]: Comment was updated to call out that we stop on `<script>` tags with no `src`! Thanks!

File: packages/angular/build/src/utils/index-file/auto-csp.ts (hunk 5 above)
Comment (aaronshim) [reply packages/angular/build/src/utils/index-file/auto-csp.ts:288]: We can look into streamlining the loader script in future bugfixes, thanks!

File: packages/angular/build/src/utils/index-file/auto-csp_spec.ts (hunk 25 above)
Comment (aaronshim) [reply packages/angular/build/src/utils/index-file/auto-csp_spec.ts:129]: Keeping the escapaing and test cases as-is for now! Thanks!

File: packages/angular/build/src/utils/index-file/auto-csp.ts [hunk 29]
```diff
@@ -0,0 +1,322 @@
...
+  'text/javascript1.1',
+  'text/javascript1.2',
+  'text/javascript1.3',
+  'text/javascript1.4',
+  'text/javascript1.5',
+  'text/jscript',
+  'text/livescript',
+  'text/x-ecmascript',
+  'text/x-javascript',
+]);
```
[the remark this answers is not in the document]
Comment (aaronshim) [reply packages/angular/build/src/utils/index-file/auto-csp.ts]: Refactored to remove all the legacy JS MIME types and only check for `text/javascript`.

**Review by dgp1130 — APPROVED** [review]

**Review by clydin — APPROVED** [review]
